Issue 4758 - Add tests for WebUI

v-cech opened the pull request:
Description:
Adding WebUI tests for bz1654238. Test for bz1654238 checks that you are able to create new entries in LDAP Browser tab.

Relates: https://github.com/389ds/389-ds-base/issues/4758

Reviewed by: @droideck

droideck commented on dirsrvtests/tests/suites/webui/ldap_browser/ldap_browser_test.py:
```diff
@@ -70,6 +70,284 @@ def test_ldap_browser_tab_visibility(topology_st, page, browser_name):
+    frame.get_by_role("button", name="Next").click()
+    frame.get_by_role("contentinfo").get_by_role("button", name="Create User").click()
+    frame.get_by_role("button", name="Finish").click()
+    frame.get_by_role("button").filter(has_text="cn=John Smith").wait_for()
```
I think a lot of the code can be encapsulated into something like `create_entry(frame, entry_type, entry_data)` function where entry_data would be a dict with rows and values.

And something similar for another test function.

droideck commented on dirsrvtests/tests/suites/webui/__init__.py:
```diff
@@ -120,3 +120,107 @@ def fin():
+        frame.get_by_role("button", name="Save row edits for row 4").click()
+        frame.get_by_role("button", name="Place row 5 in edit mode").click()
+        frame.get_by_role("textbox", name="_").fill(entry_data['user_id'])
+        frame.get_by_role("button", name="Save row edits for row 5").click()
```
I think a lot of this code can be packed with a 'for loop'.

v-cech replied:
I don't see any reason why this test is failing. I didn't modify it in any way and it was passing before. And even now when i try to run this test in a VM, it is passing without any issue. I will look into it tomorrow when the tests for new commit are finished.

Also I forgot to edit the test steps in docstring to reflect the new "load_ldap_browser_tab" function.

droideck reviewed (commented):
I think some initial and final steps also can be put in separate functions for better readability, but it's up to you. :)

Also, the CI fails for:

```
FAILED dirsrvtests/tests/suites/webui/monitoring/monitoring_test.py::test_create_credential_and_alias[firefox]
FAILED dirsrvtests/tests/suites/webui/monitoring/monitoring_test.py::test_create_credential_and_alias[chromium]
```

Is it expected and do you want to fix it separately?

droideck commented on dirsrvtests/tests/suites/webui/monitoring/monitoring_test.py:
```diff
@@ -227,6 +231,59 @@ def test_logging_visibility(topology_st, page, browser_name):
+    frame.get_by_role('button', name='Save').click()
+    assert frame.get_by_role("gridcell", name="alias.test").is_visible()
+
+
```
Regarding monitoring, I see these changes as part of your PR... May be the "sneaked in" during rebase? :)

v-cech replied:
Oh... I totally forgot that i have this test in this commit also :D For some reason i thought that it is already merged. Sorry for that.

But as i said, on my VM this failing test is passing. So i don't know what's the issue. I will remove it from this commit and try to resolve it.

droideck commented on dirsrvtests/tests/suites/webui/__init__.py:
```diff
@@ -120,3 +120,113 @@ def fin():
+                frame.get_by_role("textbox", name="_").fill(entry_data['user_id'])
+            elif row == 5:
+                frame.get_by_role("textbox", name="_").fill(entry_data['user_id'])
+            frame.get_by_role("button", name=f"Save row edits for row {row}").click()
```
I still feel that the function is just too redundant.

In my initial suggestion, I was thinking about something like that in `entry_data`:

```
entry_data = [
    {
        "type": "User",
        "name": "John Smith",
        "attrs": {
            "uid": "1204",
            "home_directory": "/user/jsmith",
            "gid": "1204",
            # etc.
            # ...
        }
    },
    {
        "type": "Group",
        "name": "testgroup"
    },
   # etc.
   # ...
    }
]
```

This way, you can get rid of all of the redundant lines and unnecessary `if`s.

v-cech replied:
Yeah, i had feeling that this is not the right approach, but i was not able to figure out something else. Thank you for guidance. Next iteration should be better.

droideck replied:
What I mean is that instead of many `elif entry_type == 'group':` blocks, you can remove the duplication and use the data from the `entry_data` dict.

For example, this code is repeated over and over for every element (more or less):

```
    elif entry_type == 'group':
        frame.get_by_role("tabpanel", name="Tree View").get_by_role("button", name="Actions").click()
        frame.get_by_role("menuitem", name="New ...").click()
        frame.get_by_label("Create a new Group").check()
        frame.get_by_role("button", name="Next").click()
        ...
        frame.get_by_role("button", name="Next").click()
        frame.get_by_role("button", name="Create", exact=True).click()
        frame.get_by_role("button", name="Finish").click()
        frame.get_by_role("button").filter(has_text=f"cn={entry_data['group_name']}").wait_for()

    elif entry_type == 'ou':
        frame.get_by_role("tabpanel", name="Tree View").get_by_role("button", name="Actions").click()
        frame.get_by_role("menuitem", name="New ...").click()
        frame.get_by_label("Create a new Organizational Unit").check()
        frame.get_by_role("button", name="Next").click()
        ...
        frame.get_by_role("button", name="Next").click()
        frame.get_by_role("button", name="Create", exact=True).click()
        frame.get_by_role("button", name="Finish").click()
        frame.get_by_role("button").filter(has_text=f"ou={entry_data['ou_name']}").wait_for()
```

So it'll be much cleaner to have the lines written only once and use the data from the entry_data dict in the places where needed (i.e. `frame.get_by_label(f"Create a new {entry_data['type']}").check()`)

v-cech replied:
So i've managed to trim some lines from each elif block. The rest of the elif blocks is unique for each entry, so i don't think i can reduce it any further. However, if you can think of anything I'd be happy to hear further comments.

@droideck

droideck replied:
I think it's possible to use  the same technic (with probably, minor modifications) that you used here:

```
for row, value in enumerate(entry_data.values()):
            if row > 5:
                break
            frame.get_by_role("button", name=f"Place row {row} in edit mode").click()
            frame.get_by_role("textbox", name="_").fill(value)
            frame.get_by_role("button", name=f"Save row edits for row {row}").click()
```

But use it generally across all objects, so you'll have no need for many `ifs` and just depend on the entry_data structure while filling in the data. Because essentially, it's repeated through all of the types (you may need to modify how entry_data is structured for checkboxes, though).

For example, for custom entry, it's very close to the example above:

```
    elif entry_type == 'custom Entry':
        frame.get_by_role("checkbox", name="Select row 0").check()
        frame.get_by_role("button", name="Next", exact=True).click()
        frame.get_by_role("checkbox", name="Select row 1").check()
        frame.get_by_role("button", name="Next", exact=True).click()
        frame.get_by_role("button", name="Place row 0 in edit mode").click()
        frame.get_by_role("textbox", name="_").fill(entry_data['uid'])
        frame.get_by_role("button", name="Save row edits for row 0").click()
        frame.get_by_role("button", name="Place row 1 in edit mode").click()
        frame.get_by_role("textbox", name="_").fill(entry_data['entry_name'])
        frame.get_by_role("button", name="Save row edits for row 1").click()
```

But okay, we spent a lot of time on this one, and it's better to progress to the next task. You can always return to this one and refactor if you'd like. :)

Still, the code looks good, and the tests pass. Ack from me.
I'll leave it for @bsimonova or @vashirov to merge if they are okay.

v-cech committed:
Issue 4758 - Add tests for WebUI

Description:
Adding WebUI tests for bz1654238. Test for bz1654238 checks that you are able to create new entries in LDAP Browser tab.

Relates: https://github.com/389ds/389-ds-base/issues/4758

Reviewed by: @droideck (Thanks!)

bsimonova commented on dirsrvtests/tests/suites/webui/ldap_browser/ldap_browser_test.py:
```diff
@@ -70,6 +87,220 @@ def test_ldap_browser_tab_visibility(topology_st, page, browser_name):
+         3. Call create_entry function to create new user.
+         4. Check that new user is successfully created.
+         5. Click on newly created user.
+         6. Call delete_entry function to delete ou.
```
Just a small nitpick 'Call delete_entry function to delete user' :)
Otherwise LGTM and tests are passing.
